BraintreeCore/src/main/java/com/braintreepayments/api/AnalyticsDatabase.kt

```diff
@@ -23,15 +25,24 @@ internal abstract class AnalyticsDatabase : RoomDatabase() {
         @Volatile
         private var INSTANCE: AnalyticsDatabase? = null
 
+        private const val OLD_VERSION = 2
+        private const val NEW_VERSION = 3
+
         // Ref: https://developer.android.com/codelabs/android-room-with-a-view-kotlin#7
         @JvmStatic
         fun getInstance(context: Context): AnalyticsDatabase =
             INSTANCE ?: synchronized(this) {
+                val MIGRATION_2_3 = object : Migration(OLD_VERSION, NEW_VERSION) {
+                    override fun migrate(database: SupportSQLiteDatabase) {
+                        database.execSQL("ALTER TABLE analytics_event ADD COLUMN link_type TEXT")
+                    }
+                }
+
                 val instance = Room.databaseBuilder(
                     context.applicationContext,
                     AnalyticsDatabase::class.java,
                     "analytics_database"
-                ).build()
+                ).addMigrations(MIGRATION_2_3).build()
```

Comment:
We didn't have a migration before this, but here too, I'd think we'd want to have a 1-2, 2-3 migration. Am I missing something? 🤔

Reply:
In this specific case (migrating the database from version 2 to version 3), a manual migration was necessary, as outlined in the [documentation](https://developer.android.com/training/data-storage/room/migrating-db-versions#manual). That’s why we don't have any records of past migrations. The exception we resolved with this approach is as follows:

java.lang.IllegalStateException: Migration didn't properly handle: analytics_event(com.braintreepayments.api.AnalyticsEvent).
 Expected:
TableInfo{name='analytics_event', columns={name=Column{name='name', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=0, defaultValue='null'}, link_type=Column{name='link_type', type='TEXT', affinity='2', notNull=false, primaryKeyPosition=0, defaultValue='null'}, _id=Column{name='_id', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=1, defaultValue='null'}, paypal_context_id=Column{name='paypal_context_id', type='TEXT', affinity='2', notNull=false, primaryKeyPosition=0, defaultValue='null'}, timestamp=Column{name='timestamp', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='null'}}, foreignKeys=[], indices=[]}
 Found:
TableInfo{name='analytics_event', columns={name=Column{name='name', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=0, defaultValue='null'}, _id=Column{name='_id', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=1, defaultValue='null'}, paypal_context_id=Column{name='paypal_context_id', type='TEXT', affinity='2', notNull=false, primaryKeyPosition=0, defaultValue='null'}, timestamp=Column{name='timestamp', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='null'}}, foreignKeys=[], indices=[]}


It appears that the migration does not recognize that a new column was added to the table. In the manual migration, we are indicating that the column `link_type` was added to the table `analytics_event`.